File: src/main/java/de/solidblocks/examples/javaspringwuphf/ExampleApplication.java
package de.solidblocks.examples.javaspringwuphf;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;

@SpringBootApplication
public class ExampleApplication {

    /** Profile that keeps the visits in postgres, and the one that keeps them in memory. */
    static final String DATABASE_PROFILE = "database";
    static final String MEMORY_PROFILE = "memory";

    /**
     * The environment variable a blcks link to a postgresql-standalone service named {@code
     * database1} sets. Its presence is what decides where the visits are kept: linked, they go to
     * the database; unlinked - running the archive by hand, say - they stay in the process.
     */
    static final String HOST_VARIABLE = "BLCKS_DATABASE1_HOST";

    public static void main(String[] args) {
        SpringApplication application = new SpringApplication(ExampleApplication.class);
        application.setAdditionalProfiles(profile(System.getenv(HOST_VARIABLE)));
        application.run(args);
    }

    static String profile(String host) {
        return host == null || host.isBlank() ? MEMORY_PROFILE : DATABASE_PROFILE;
    }

    /**
     * The environment of the process, read once at start-up.
     *
     * <p>The map is given to the constructor here rather than read in it. A {@code Map<String,
     * String>} parameter of a bean is a parameter that spring fills with all beans of type {@code
     * String} by their name, and the environment would then be empty without an error.
     */
    @Bean
    ProcessEnvironment processEnvironment() {
        return new ProcessEnvironment(System.getenv());
    }
}

File: src/main/kotlin/de/solidblocks/examples/javaspringwuphf/ExampleApplication.kt
package de.solidblocks.examples.javaspringwuphf

import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.runApplication
import org.springframework.context.annotation.Bean

@SpringBootApplication
class ExampleApplication {
    @Bean fun processEnvironment() = ProcessEnvironment(System.getenv())

    companion object {
        const val DATABASE_PROFILE = "database"
        const val MEMORY_PROFILE = "memory"

        const val HOST_VARIABLE = "BLCKS_DATABASE1_HOST"

        fun profile(host: String?) = if (host.isNullOrBlank()) MEMORY_PROFILE else DATABASE_PROFILE
    }
}

fun main(args: Array<String>) {
    runApplication<ExampleApplication>(*args) {
        setAdditionalProfiles(ExampleApplication.profile(System.getenv(ExampleApplication.HOST_VARIABLE)))
    }
}
